nucypher/utilities/networking.py

```diff
@@ -15,42 +15,167 @@
  along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
 """
 
-import click
+
+import random
 import requests
+from ipaddress import ip_address
+from requests.exceptions import RequestException, HTTPError
+from typing import Union
 
-from nucypher.cli.literature import CONFIRM_URSULA_IPV4_ADDRESS, COLLECT_URSULA_IPV4_ADDRESS, \
-    FORCE_DETECT_URSULA_IP_WARNING
-from nucypher.cli.types import IPV4_ADDRESS
+from nucypher.config.storages import LocalFileBasedNodeStorage
+from nucypher.acumen.perception import FleetSensor
+from nucypher.blockchain.eth.registry import BaseContractRegistry, InMemoryContractRegistry
+from nucypher.network.middleware import RestMiddleware, NucypherMiddlewareClient
+from nucypher.utilities.logging import Logger
 
 
 class UnknownIPAddress(RuntimeError):
     pass
 
 
-def get_external_ip_from_centralized_source() -> str:
-    ip_request = requests.get('https://ifconfig.me/')
-    if ip_request.status_code == 200:
-        return ip_request.text
-    raise UnknownIPAddress(f"There was an error determining the IP address automatically. "
-                           f"(status code {ip_request.status_code})")
+class InvalidWorkerIP(RuntimeError):
+    """Raised when an Ursula is using an invalid IP address for it's server."""
+
+
+RequestErrors = (
+    # https://requests.readthedocs.io/en/latest/user/quickstart/#errors-and-exceptions
+    ConnectionError,
+    TimeoutError,
+    RequestException,
+    HTTPError
+)
+
+RESERVED_IP_ADDRESSES = (
+    '0.0.0.0',
+    '127.0.0.1',
+    '1.2.3.4'
+)
+
+IP_DETECTION_LOGGER = Logger('external-ip-detection')
 
 
-def determine_external_ip_address(emitter, force: bool = False) -> str:
+
+def validate_worker_ip(worker_ip: str) -> None:
+    if worker_ip in RESERVED_IP_ADDRESSES:
+        raise InvalidWorkerIP(f'{worker_ip} is not a valid or permitted worker IP address.  '
+                              f'Verify the rest_host is set to the external IPV4 address')
+
+
+def __request(url: str, certificate=None) -> Union[str, None]:
     """
-    Attempts to automatically get the external IP from ifconfig.me
-    If the request fails, it falls back to the standard process.
+    Utility function to send a GET request to a URL returning it's
+    text content or None, suppressing all errors. Certificate is
+    needed if the remote URL source is self-signed.
     """
     try:
-        rest_host = get_external_ip_from_centralized_source()
-    except UnknownIPAddress:
-        if force:
-            raise
+        # 'None' or 'True' will verify self-signed certificates
+        response = requests.get(url, verify=certificate)
+    except RequestErrors:
+        return None
+    if response.status_code == 200:
+        return response.text
```

Comment:
What about other status codes? I guess it implicitly returns None, but a explicit value would be nice.